src/volatile_memory.rs

```diff
@@ -219,13 +235,57 @@ impl<'a> From<&'a mut [u8]> for VolatileSlice<'a, ()> {
 #[repr(C, packed)]
 struct Packed<T>(T);
 
+// Guards the ptr_guard() and ptr_guard_mut() implementations by mapping memory before the
+// operations are performed and unmapping after they are done.
+struct AsPtrGuard {
+    addr: *mut u8,
+
+    // This isn't used anymore, but it protects the slice from getting unmapped while in use.
+    // Once this goes out of scope, the memory is unmapped automatically.
+    #[cfg(all(feature = "xen", unix))]
+    _mmap_slice: MmapSlice,
+}
+
+impl AsPtrGuard {
+    #[allow(unused_variables)]
+    fn new(mmap: Option<&MmapInfo>, addr: *mut u8, len: usize, read_only: bool) -> Self {
+        #[cfg(all(feature = "xen", unix))]
+        let (addr, _mmap_slice) = {
+            let prot = if read_only {
+                libc::PROT_READ
+            } else {
+                libc::PROT_WRITE
+            };
+
+            let slice = match mmap {
+                Some(mmap) => mmap.mmap_slice(addr, prot, len),
+                None => MmapSlice::raw(addr),
+            };
+
+            (slice.addr(), slice)
+        };
+
```

Comment:
I think we can simplify a bit more here. Because really, the "guard" is `MmapSlice`: Its what unmaps the memory when it gets dropped. So this can simply be a utility function that returns a `MmapSlice` and then at the guarded callsites, we use `MmapSlice::addr` instead of `AsPtrGuard::as_ptr`.

Reply:
> I think we can simplify a bit more here. Because really, the "guard" is `MmapSlice`: Its what unmaps the memory when it gets dropped. So this can simply be a utility function that returns a `MmapSlice` and then at the guarded callsites, we use `MmapSlice::addr` instead of `AsPtrGuard::as_ptr`.

I am not sure I understood it properly. `MmapSlice` is a Xen thing, and doesn't exist otherwise. And so the call sites do need to use another abstraction, like `AsPtrGuard`. What am I missing ?